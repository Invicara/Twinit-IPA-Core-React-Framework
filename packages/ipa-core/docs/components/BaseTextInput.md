---
title: BaseTextInput
sidebar_position: 1000
---
A text input that provides basic functionality and styling but can be extended or composed to create a wide range of text input variants.

### Properties

- `className`: a className appended to the default classNames.
- `inputProps`: *(required)* Object of props related to the input, they are used to pass properties to the underlying `input` tag or to the input component provided in props (see component property below).
  - `type`: It is be passed directly to the underlying `input` or alternative `component`.
  - `value`: It is be passed directly to the underlying `input` or alternative `component`.
  - `onChange`: It is be passed directly to the underlying `input` or alternative `component`.
  - `placeholder`: It is be passed directly to the underlying `input` or alternative `component`.
  - `disabled`: It is be passed directly to the underlying `input` or alternative `component`.
  - `className`: a `className` `string` to append to the default classNames, it is one of the two ways you can style the `input`.
  - `style`: a `style` `object` passed directly to the underlying `input` or alternative `component`, it is the other way you can style the `input`.
  - `onFocusChange`: a `function` callback called with argument `true` when the underlying `input` or component gets the focus, and `false`, when it loses the focus. 
- `labelProps`: `object` of `props` related to the label, they will be passed to the underlying `ControlLabel` component. If not provided, the `ControlLabel` will not be displayed.
- `component`: An alternative component to the react `input` tag.
- `children`: A React `Node` or an `array` of `Node`  that is displayed after the `input` in the DOM.


### PropTypes
#### BaseTextInput
```jsx
{
    className: string,
    labelProps: ControlLabel.propTypes,
    inputProps: shape({
        type: string,
        value: any,
        onChange: func,
        placeholder: string,
        disabled: bool,
        className: string,
        style: object,
        onFocusChange: func,
    }).isRequired,
    component: element,
}
```

#### ControlLabel
```jsx
{
    style: object,
    text: string,
    className: string,
}
```

### Examples
You can find examples of usage in the components below :
- **`EntityModalTextInput`**: A classic example of composition, the `BaseTextInput` is used without the `component` prop but wrapped in another component to add a text overlay that responds to the input's focus.
- **`CollapsibleTextInput`**: This component uses `BaseTextInput` but changes it's core `input` component to another component with specific features when a `button` passed as `children` is clicked.
