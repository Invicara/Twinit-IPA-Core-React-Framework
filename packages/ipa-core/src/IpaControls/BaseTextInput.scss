.base-text-input {
  display: flex;
  flex-direction: column;
  input:focus {
    border: 2px solid var(--app-accent-color) !important;
    box-shadow: unset;
  }
}