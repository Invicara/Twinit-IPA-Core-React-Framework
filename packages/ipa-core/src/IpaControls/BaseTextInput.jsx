import React, { useState } from "react";
import PropTypes from "prop-types";
import { any, bool, element, func, object, shape, string } from "prop-types";
import "./BaseTextInput.scss";
import ControlLabel from "./ControlLabel";

const BaseTextInput = (props) => {
  const [isFocused, setIsFocused] = useState(false);

  const handleFocus = () => {
    setIsFocused(true);
    props.inputProps.onFocusChange?.(true);
  };

  const handleBlur = () => {
    setIsFocused(false);
    props.inputProps.onFocusChange?.(false);
  };

  const baseClassName = "base-text-input";
  let classNameModifiers = isFocused ? `${baseClassName}--is-focused` : "";

  let inputProps = {
    value: props.inputProps.value,
    type: props.inputProps.type,
    onChange: props.inputProps.onChange,
    placeholder: props.inputProps.placeholder,
    disabled: props.inputProps.disabled,
    onFocus: handleFocus,
    onBlur: handleBlur,
    className: `form-control ${props.inputProps.className}`,
    style: props.style,
  };

  return (
    <div
      className={`${baseClassName} ${props.className} ${classNameModifiers}`}
    >
      {props.labelProps && <ControlLabel {...props.labelProps} />}
      {props.component ? (
        <props.component {...inputProps} />
      ) : (
        <input {...inputProps} />
      )}
      {props.children}
    </div>
  );
};

BaseTextInput.propTypes = {
  className: string,
  labelProps: shape({ ...ControlLabel.propTypes }),
  inputProps: shape({
    type: string,
    value: any,
    onChange: func,
    placeholder: string,
    disabled: bool,
    className: string,
    style: object,
    onFocusChange: func,
    onBlur: func,
  }).isRequired,
  component: element,
};

export default BaseTextInput;
