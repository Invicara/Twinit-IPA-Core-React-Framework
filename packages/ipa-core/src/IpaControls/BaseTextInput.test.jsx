import renderer from 'react-test-renderer';
import BaseTextInput from './BaseTextInput';
import React from 'react';

const homepageErrors = console.error.bind(console.error)
beforeAll(() => {
    console.error = errormessage => {
        /*
          if error is a proptype error and includes the following string: `Warning: Failed prop type:`
          suppress the error and don't show it
          if it is not a proptype error, we show it
        */
        const suppressedErrors = errormessage
            .toString()
            .includes("Warning: Failed prop type:")

        !suppressedErrors && homepageErrors(errormessage)
    }
})
afterAll(() => {
    console.error = homepageErrors
})


it('change la classe lorsqu\'on le survole', () => {
  const component = renderer.create(
    <BaseTextInput inputProps={{
        value: "",
        onChange: () => {console.log("pwett")},
        placeholder: "placeholder",
        className: "test-base-text-input",
        onFocusChange: () => {console.log("focus")},
    }}/>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});