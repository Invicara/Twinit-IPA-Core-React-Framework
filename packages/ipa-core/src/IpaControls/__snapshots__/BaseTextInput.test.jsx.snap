// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`change la classe lorsqu'on le survole 1`] = `
<div
  className="base-text-input undefined "
>
  <input
    className="form-control test-base-text-input"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="placeholder"
    value=""
  />
</div>
`;
